### DISCUSSION TITLE
Silly question about "_".

### DISCUSSION THREAD

**Discussion by marcovc** [hsutter/cppfront#747] [Q&A] [answered]
Hi!

I always thought that we use "auto" in (standard) c++ in the following piece of code:

```c++
auto i = 3;
```

just because doing just

```c++
i = 3;
```

would be "too disruptive". But now I see that even when we are free from backward compatibility, there it is "auto" again, this time with a new symbol "_". I was just curious, why can't we get always away without specifying the type at all when we want it to be automatically inferred, on in other words, why can't we denote "auto" as ""?

**Comment by msadeqhe** [comment, marked as answer]
That's a good question. Actually `_` is optional except for template parameters:

```cpp
fnc1: (x: _) -> _ = x + 1;
fnc2: (x)           x + 1;

var1: _ = 10;
var2:   = 10;

// Type template parameter
cls1: <T: type> type = { }
cls2: <T>       type = { }

// Non-type template parameter
cls3: <T: _> type = { }
```

Because _"type template parameters"_ are the default type. If we want _"non-type template parameters"_, we have to explicitly write it.

**Reply by marcovc** [reply]
I see, thanks! 

Somehow related, when we don't specificy "_", why do we still need sometimes to include the type specifier prefix ":" ? For example, why do we do:

```cpp2
var2:   = 10;
```

instead of

```cpp2
var2 = 10;
```

but we do

```cpp2
fnc2: (x)         = { return x + 1; }
```

(which makes sense) instead of

```cpp2
fnc2: (x:)         = { return x + 1; }
```
?

**Reply by MaxSagebaum** [reply]
```
var2: = 10;
```
creates a new variable with the name `var2`.

```
var2 = 10;
```
assigns a value to an existing variable.

**Reply by marcovc** [reply]
Oh I see, it would be impossible to distinguish those two cases if we don't include the type specifier prefix ":". Thanks.

**Reply by hsutter** [reply]
Yes, in Cpp2 `:` always-and-only means "declaring a new thing":

- "always" = every declaration uses a `:`
- "only" = nothing else in the language uses `:`

Even when I decided to allow naming loops with `outer: for /*etc.*/` I view it as declaring a named loop (and unnamed loops/statements are just 'undeclared' things). I don't think that stretched it too far and that the above strict "always and only" still holds reasonably strictly.

**Reply by jcanizales** [reply]
Function parameters aren't "declared"?

**Reply by gregmarr** [reply]
I guess you can say that the "always" part of "always-and-only" has been relaxed a little bit with the ability to leave off `: _` for parameters of deduced types.

**Reply by JohelEGP** [reply]
I think that's [defaults](https://github.com/hsutter/cppfront/wiki/Design-note%3A-Defaults-are-one-way-to-say-the-same-thing) at play.

**Reply by gregmarr** [reply]
Yes, that's what I was referring to.  So for those not familiar with the defaults, it's more like "Every declaration uses a `:`, unless it has been defaulted."

**Reply by jcanizales** [reply]
It we ever have default argument values, they might be declared like `x = 10` for deduced type then.

**Reply by JohelEGP** [reply]
No, that's an assignment.
The default for parameter declarations omits `: type = value`.
That's why statement parameters need the `:`.
```Cpp2
(x := 0) { … }
```

**Reply by jcanizales** [reply]
Oh we already have default argument values? I thought they were on hold pending to see if function overloading was a good enough substitute.

**Reply by JohelEGP** [reply]
Only for parameters of statements (see <https://github.com/hsutter/cppfront/wiki/Design-note%3A-Defaults-are-one-way-to-say-the-same-thing>).

**Reply by jcanizales** [reply]
Oh but that note shows as example:

```ruby
f:(x: int = init) = { ... }
```

Is it an "in the works" thing? (I hope the asymmetry with parameterized statements is only temporary)

**Reply by JohelEGP** [reply]
See #75.

**Comment by SebastianTroy** [comment]
I think it would become a cause of bugs. Imagine in a new scope you want a variable with a common name like i, so you use i = 0, and it actually sets the value in an outer scope. At least with i: = 0 the cpp2 transpiler can warn about a variable name being reused, and the person maintaining the code can see that a new variable was intended, rather than have to guess that's what the author of the code meant


On 12 October 2023 09:04:17 Marco Correia ***@***.***> wrote:

Oh I see, it would be impossible to distinguish those two cases if we don't include the type specifier prefix ":". Thanks.

**Comment by ntrel** [comment]
@msadeqhe 
> That's a good question. Actually `_` is optional except for template parameters:
> 
> ```c++
> fnc1: (x: _) -> _ = { return x + 1; }
> fnc2: (x)         = { return x + 1; }

Actually `fnc2` is an error, you have to use ~~`= expression;` to get return type inference:~~

~~fnc2: (x)         = x + 1;~~
```c++
fnc3: (x)           x + 1;
```
~~BTW given we now have the `fnc3` form, I don't see why we need the `fnc2` form with implicit `return`. I would get rid of that and just have `= statement`.~~

**Reply by JohelEGP** [reply]
> the `fnc2` form with implicit `return`

There's no implicit `return`.
These are equivalent:
```Cpp2
fnc2: (x: _) -> void = { x + 1; }
fnc2: (x)         = x + 1;
```
It's an error in Cpp2 because you didn't discard the value.
You get it with `-Werror=unused-result -Werror=unused-value` on Clang and GCC: <https://cpp2.godbolt.org/z/hEMjEMz1r>.
Or maybe it should be guaranteed by Cppfront regardless of Cpp1 compiler, e.g., lowered `CPP2_VOID(x + 1)`.

**Reply by msadeqhe** [reply]
Thanks. I fixed it.

**Reply by ntrel** [reply]
@JohelEGP thanks, I see.
